src/autonomous/AutonomousTask.cpp

```diff
@@ -0,0 +1,57 @@
+#include "AutonomousTask.h"
+
+#include "../Constants.h"
+#include "../Globals.h"
+#include "../commands/DriveToWaypointCommand.h"
+#include "../world_interface/world_interface.h"
+
+namespace autonomous {
+
+AutonomousTask::AutonomousTask(){};
+
+AutonomousTask::~AutonomousTask() {
+	if (_autonomous_task_thread.joinable()) {
+		_autonomous_task_thread.join();
+	}
+}
+
+void AutonomousTask::start(const navtypes::pose_t& waypointCoords) {
+	setWaypoint(waypointCoords);
+	_kill_called = false;
+	_autonomous_task_thread = std::thread(&autonomous::AutonomousTask::navigate, this);
+}
+
+void AutonomousTask::navigate() {
+	commands::DriveToWaypointCommand cmd(_waypoint_coords, THETA_KP, DRIVE_VEL, SLOW_DRIVE_VEL,
+										 DONE_THRESHOLD, CLOSE_TO_TARGET_DUR_VAL);
+
+	while (!cmd.isDone()) {
+		auto latestGPS = robot::readGPS();
+		if (latestGPS.isFresh(2000ms)) {
+			cmd.setState(latestGPS, robot::types::dataclock::now());
+			commands::command_t output = cmd.getOutput();
+			robot::setCmdVel(output.thetaVel, output.xVel);
+		}
+		std::unique_lock autonomousTaskLock(_autonomous_task_mutex);
+		// Wait 1000ms or return out of while loop if kill called
+		if (_autonomous_task_cv.wait_until(autonomousTaskLock,
+										   std::chrono::system_clock::now() + 1000ms,
+										   [&] { return _kill_called; })) {
+			return;
+		}
```

Comment:
This sleeps for 1000ms, so it assumes that the rest of the code runs in 0 time. That's not true, so we'll get an unsteady update rate.

Instead, initialize a `sleepUntil` time outside of the loop to the current time, and increment it by 1000ms and sleep every time. This will make it a more consistent update rate.

Reply:
Also, we should use `std::chrono::steady_clock` instead of `std::chrono::system_clock`. Read about the difference [here](https://stackoverflow.com/questions/31552193/difference-between-steady-clock-vs-system-clock).